apps/src/blockly/addons/cdoUtils.js

```diff
@@ -106,9 +109,167 @@ export function getUserTheme(themeOption) {
+    (experiments.isEnabled(experiments.BLOCKLY_JSON) || Math.random() <= 0.05)
+  ) {
+    testJsonSerialization(workspace);
+  }
   return Blockly.Xml.domToText(Blockly.Xml.blockSpaceToDom(workspace));
-  // After supporting JSON block sources, change to:
-  // return JSON.stringify(Blockly.serialization.workspaces.save(workspace));
+}
+
+function testJsonSerialization(workspace) {
+  const FIREHOSE_STUDY = 'blockly-json';
+  const FIREHOSE_EVENT = 'block-differences';
+  Blockly.Events.disable();
+
+  // Create an array of blocks based on JSON serialization of the current workspace.
+  const blockJson = Blockly.serialization.workspaces.save(workspace);
+  const tempJsonWorkspace = new Blockly.Workspace();
+  Blockly.serialization.workspaces.load(blockJson, tempJsonWorkspace);
+  const jsonBlocks = tempJsonWorkspace.getAllBlocks();
+
+  // Create an array of blocks based on the XML encoding of the current workspace.
+  const blockXml = Blockly.Xml.blockSpaceToDom(workspace);
+  const tempXmlWorkspace = new Blockly.Workspace();
+  Blockly.Xml.domToWorkspace(blockXml, tempXmlWorkspace);
+  const xmlBlocks = tempXmlWorkspace.getAllBlocks();
+
+  // compareBlockArrays returns an array of differences found.
+  const differences = compareBlockArrays(xmlBlocks, jsonBlocks);
+  if (differences.length > 0) {
+    // Log a record to Firehose/Redshift.
+    const recordData = {
+      projectUrl: dashboard.project.getShareUrl(),
+      blocklyVersion: Blockly.blockly_.VERSION,
+      differences: differences,
+    };
+    firehoseClient.putRecord(
+      {
+        study: FIREHOSE_STUDY,
+        event: FIREHOSE_EVENT,
+        data_json: JSON.stringify(recordData),
+      },
+      {
+        alwaysPut: true, // Allows logging from development environments
+        includeUserId: true,
+      }
+    );
+  }
+  if (experiments.isEnabled(experiments.BLOCKLY_JSON)) {
+    // Avoid logging to the console for typical users.
+    console.log({differences: differences});
+  }
+}
+
+// Used to find differences between blocks created from xml and json sources.
+// Called when the app is initialized and each time the app runs.
+// Compares two block arrays and returns an array of differences between them.
+export function compareBlockArrays(xmlBlocks, jsonBlocks) {
+  // compareValues() will be called recursively so we need to keep track of objects
+  // that it has already compared, to prevent infinite loops caused by circular
+  // references.
+  const visited = new Set();
+  // Some block properties are expected to hold different values, or refer to blocks already in the array.
+  const keysToSkip = [
+    'parentBlock_', // a circular reference to another block in the array
+    'sourceBlock_', // a circular reference to another block in the array
+    'workspace', // a reference to the unique temporary workspaces used to create the two block arrays
+  ];
+
+  function compareValues(arg1, arg2, path = '', differences = []) {
```

Comment:
Does this become a generic "compare objects" function if we pass in `keysToSkip` as an argument? Might be nice to isolate this as a utility.

Reply:
Cool idea @bencodeorg

Reply:
Going to call this out of scope for now. :)